### PR TITLE
Prisma instrumentation

### PR DESCRIPTION
**Pull request by bizob2828**
## Proposed Release Notes
 * Added support for Prisma.
   * Captures spans for queries. It names them based on the model and action.(i.e. Datastore/statement/Prisma/user/create)
   * Captures database metrics for queries.
   * Provides connection between application and database server via service maps. 

Huge shoutout to @osmanmrtacar for the original contribution 🙏🏻 

## Links
 * This builds on top of https://github.com/newrelic/node-newrelic/pull/1512. 
 * We also plan to write versioned tests in a different PR.

## Details
You can see here the span names.  It also connects the app with db.  

screenshot 2023-02-27 at 5 01 23 PM

When using `prisma.$executeRaw` we will name the spans as `Datastore/statement/Prisma/$executeRaw(<sql cmd>)/<database name>`

screenshot 2023-02-27 at 5 04 07 PM

I also had to extend the datastore shim which I plan on writing unit tests for to take advantage of the `inContext` method to set the params for the database host/port/db.

Lastly, the original PR instrumented `_request` but `_executeRequest` seems more appropriate as it's a higher level function that gets called

### CODE REVIEW

File: lib/instrumentation/@prisma/client.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+const connection = Symbol('nrPrismaConnection')
+const url = require('url')
+const logger = require('../../logger').child({ component: 'prisma' })
+const parseSql = require('../../db/query-parsers/sql')
+
+/**
+ * Parses a connection string. Most database engines in prisma are SQL and all
+ * have similar engine strings.
+ *
+ * **Note**: This will not parse mongodb or sql server, instead will log a warning
```
Comment (jordigh) [inline lib/instrumentation/@prisma/client.js:17]: Hm, still kinda surprised to hear that. Reading the mongo docs, it uses standard URI for its connexion strings.

File: lib/instrumentation/@prisma/client.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  let parameters = {}
+  try {
+    const parsedUrl = url.parse(connectionUrl)
+    parameters = {
+      host: parsedUrl.hostname,
+      port: parsedUrl.port,
+      database_name: parsedUrl.path && decodeURIComponent(parsedUrl.path.split('/')[0])
+    }
+  } catch (err) {
+    logger.warn('Failed to parse connection string for %s: %s', provider, err.message)
```
Comment (jordigh) [inline lib/instrumentation/@prisma/client.js:33]: Is there a reason to prefer format strings over template literals?

File: lib/shim/datastore-shim.js [hunk 3]
```diff
@@ -693,6 +694,7 @@ function _recordQuery(suffix, nodule, properties, querySpec) {
       promise: 'promise' in queryDesc ? queryDesc.promise : null,
       internal: 'internal' in queryDesc ? queryDesc.internal : true,
       opaque: 'opaque' in queryDesc ? queryDesc.opaque : false,
+      inContext: 'inContext' in queryDesc ? queryDesc.inContext : null,
```
Comment (jordigh) [inline lib/shim/datastore-shim.js:697]: Neat, didn't know segment descriptors could have this attribute.

**Review by jordigh — COMMENTED** [review]
Seems alright, should probably just keep the one added symbol with all of the others we already have.

File: lib/instrumentation/@prisma/client.js (hunk 1 above)
Comment (bizob2828) [reply lib/instrumentation/@prisma/client.js:17]: you're right. not sure why I thought that

File: lib/instrumentation/@prisma/client.js (hunk 2 above)
Comment (bizob2828) [reply lib/instrumentation/@prisma/client.js:33]: not really.  template strings are newer and sometimes I'm just used to doing this when logging

File: lib/instrumentation/@prisma/client.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      return {
+        promise: true,
+        query: retrieveQuery(args),
+        /**
+         * Adds the relevant host, port, database_name parameters
+         * to the active segment
+         */
+        inContext() {
+          if (!client[connection]) {
+            client._engine.getConfig().then((prismaConfig) => {
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/instrumentation/@prisma/client.js:99]: yea I tried to instrument this function but I couldn't seem to get at it at time so this was the compromise. it should only call this promise once

File: lib/shim/datastore-shim.js [hunk 5]
```diff
@@ -5,7 +5,7 @@
 
 'use strict'
 
-/* eslint sonarjs/cognitive-complexity: ["error", 52] -- TODO: https://issues.newrelic.com/browse/NEWRELIC-5252 */
+/* eslint sonarjs/cognitive-complexity: ["error", 57] -- TODO: https://issues.newrelic.com/browse/NEWRELIC-5252 */
```
Comment (bizob2828) [inline lib/shim/datastore-shim.js:8]: revert

File: lib/shim/datastore-shim.js (hunk 3 above)
Comment (bizob2828) [reply lib/shim/datastore-shim.js:697]: yea it's used in message-shim. I didn't add support for this in all cases in here as this is the only use case atm

**Review by jordigh — APPROVED** [review]

**Pull request merged**